File: apps/android/shared/src/main/java/com/eatfair/shared/model/driver/Driver.kt
package com.eatfair.shared.model.driver

/**
 * Driver model for driver management (API-based)
 * Used for driver profile, onboarding, and status tracking
 */
data class Driver(
    val id: Int = 0,  // Backend returns Integer
    val driverId: String = "",  // Business ID like "DRV-00002"
    val name: String = "",
    val email: String = "",
    val phone: String = "",
    val profileImageUrl: String? = null,
    val dateOfBirth: Long? = null,
    val ssn4: String? = null,  // Last 4 digits only for security

    // Address
    val address: DriverAddress? = null,

    // Driver's License
    val driversLicense: DriversLicense? = null,

    // Vehicle
    val vehicle: VehicleInfo? = null,
    val vehicleType: String = "car",  // "car", "bike", "scooter", "motorcycle"
    val licensePlate: String = "",

    // Insurance
    val insurance: InsuranceInfo? = null,

    // Bank Account
    val bankAccount: BankAccountInfo? = null,

    // Status - matches backend field name
    val isOnline: Boolean = false,
    val isApproved: Boolean = false,
    val status: String = "pending",  // "pending", "active", "suspended", "deactivated"
    val backgroundCheckStatus: String? = null,  // "pending", "passed", "failed"
    val backgroundCheckDate: Long? = null,

    // Current Location
    val currentLatitude: Double = 0.0,
    val currentLongitude: Double = 0.0,
    val lastActive: Long? = null,
    val currentSessionId: String? = null,

    // Stats (CRITICAL - was missing in Android)
    val stats: DriverStats? = null,

    // Preferences
    val preferences: DriverPreferences? = null,

    // Timestamps
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long? = null,
    val onboardingCompletedAt: Long? = null
)

data class DriverAddress(
    val street: String = "",
    val unit: String? = null,
    val city: String = "",
    val state: String = "",
    val zipCode: String = "",
    val country: String = "US"
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "street" to street,
        "unit" to unit,
        "city" to city,
        "state" to state,
        "zipCode" to zipCode,
        "country" to country
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): DriverAddress {
            return DriverAddress(
                street = data["street"] as? String ?: "",
                unit = data["unit"] as? String,
                city = data["city"] as? String ?: "",
                state = data["state"] as? String ?: "",
                zipCode = data["zipCode"] as? String ?: "",
                country = data["country"] as? String ?: "US"
            )
        }
    }
}

data class DriversLicense(
    val licenseNumber: String = "",
    val state: String = "",  // Issuing state
    val expirationDate: Long = 0,
    val licenseClass: String = "C",  // "A", "B", "C", "D", "M"
    val frontImageUrl: String? = null,
    val backImageUrl: String? = null,
    val isVerified: Boolean = false,
    val verifiedAt: Long? = null
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "licenseNumber" to licenseNumber,
        "state" to state,
        "expirationDate" to expirationDate,
        "licenseClass" to licenseClass,
        "frontImageUrl" to frontImageUrl,
        "backImageUrl" to backImageUrl,
        "isVerified" to isVerified,
        "verifiedAt" to verifiedAt
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): DriversLicense {
            return DriversLicense(
                licenseNumber = data["licenseNumber"] as? String ?: "",
                state = data["state"] as? String ?: "",
                expirationDate = (data["expirationDate"] as? Number)?.toLong() ?: 0,
                licenseClass = data["licenseClass"] as? String ?: "C",
                frontImageUrl = data["frontImageUrl"] as? String,
                backImageUrl = data["backImageUrl"] as? String,
                isVerified = data["isVerified"] as? Boolean ?: false,
                verifiedAt = (data["verifiedAt"] as? Number)?.toLong()
            )
        }
    }
}

data class VehicleInfo(
    val make: String = "",  // Toyota, Honda, Ford
    val model: String = "",  // Camry, Civic, F-150
    val year: Int = 0,
    val color: String = "",
    val licensePlate: String = "",
    val state: String = "",  // License plate state
    val vehicleType: String = "sedan",  // "sedan", "suv", "truck", "motorcycle", "bicycle", "scooter"
    val frontImageUrl: String? = null,
    val sideImageUrl: String? = null,
    val backImageUrl: String? = null,
    val registrationNumber: String? = null,
    val registrationExpirationDate: Long? = null,
    val registrationImageUrl: String? = null,
    val isVerified: Boolean = false
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "make" to make,
        "model" to model,
        "year" to year,
        "color" to color,
        "licensePlate" to licensePlate,
        "state" to state,
        "vehicleType" to vehicleType,
        "frontImageUrl" to frontImageUrl,
        "sideImageUrl" to sideImageUrl,
        "backImageUrl" to backImageUrl,
        "registrationNumber" to registrationNumber,
        "registrationExpirationDate" to registrationExpirationDate,
        "registrationImageUrl" to registrationImageUrl,
        "isVerified" to isVerified
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): VehicleInfo {
            return VehicleInfo(
                make = data["make"] as? String ?: "",
                model = data["model"] as? String ?: "",
                year = (data["year"] as? Number)?.toInt() ?: 0,
                color = data["color"] as? String ?: "",
                licensePlate = data["licensePlate"] as? String ?: "",
                state = data["state"] as? String ?: "",
                vehicleType = data["vehicleType"] as? String ?: "sedan",
                frontImageUrl = data["frontImageUrl"] as? String,
                sideImageUrl = data["sideImageUrl"] as? String,
                backImageUrl = data["backImageUrl"] as? String,
                registrationNumber = data["registrationNumber"] as? String,
                registrationExpirationDate = (data["registrationExpirationDate"] as? Number)?.toLong(),
                registrationImageUrl = data["registrationImageUrl"] as? String,
                isVerified = data["isVerified"] as? Boolean ?: false
            )
        }
    }
}

data class InsuranceInfo(
    val provider: String = "",  // State Farm, Geico, Progressive
    val policyNumber: String = "",
    val expirationDate: Long = 0,
    val coverageType: String = "liability",  // "liability", "full_coverage"
    val insuranceCardImageUrl: String? = null,
    val isVerified: Boolean = false
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "provider" to provider,
        "policyNumber" to policyNumber,
        "expirationDate" to expirationDate,
        "coverageType" to coverageType,
        "insuranceCardImageUrl" to insuranceCardImageUrl,
        "isVerified" to isVerified
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): InsuranceInfo {
            return InsuranceInfo(
                provider = data["provider"] as? String ?: "",
                policyNumber = data["policyNumber"] as? String ?: "",
                expirationDate = (data["expirationDate"] as? Number)?.toLong() ?: 0,
                coverageType = data["coverageType"] as? String ?: "liability",
                insuranceCardImageUrl = data["insuranceCardImageUrl"] as? String,
                isVerified = data["isVerified"] as? Boolean ?: false
            )
        }
    }
}

data class BankAccountInfo(
    val bankName: String = "",
    val accountHolderName: String = "",
    val accountType: String = "checking",  // "checking", "savings"
    val routingNumber: String = "",  // Masked for security
    val accountNumberLast4: String = "",  // Only last 4 digits stored
    val isVerified: Boolean = false,
    val stripeConnectId: String? = null  // For Stripe Connect payouts
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "bankName" to bankName,
        "accountHolderName" to accountHolderName,
        "accountType" to accountType,
        "routingNumber" to routingNumber,
        "accountNumberLast4" to accountNumberLast4,
        "isVerified" to isVerified,
        "stripeConnectId" to stripeConnectId
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): BankAccountInfo {
            return BankAccountInfo(
                bankName = data["bankName"] as? String ?: "",
                accountHolderName = data["accountHolderName"] as? String ?: "",
                accountType = data["accountType"] as? String ?: "checking",
                routingNumber = data["routingNumber"] as? String ?: "",
                accountNumberLast4 = data["accountNumberLast4"] as? String ?: "",
                isVerified = data["isVerified"] as? Boolean ?: false,
                stripeConnectId = data["stripeConnectId"] as? String
            )
        }
    }
}

/**
 * CRITICAL: Driver stats - was completely missing in Android
 * This enables proper earnings tracking, ratings, and performance metrics
 */
data class DriverStats(
    val rating: Double = 5.0,  // 1.0-5.0
    val totalDeliveries: Int = 0,
    val completedDeliveries: Int = 0,
    val cancelledDeliveries: Int = 0,
    val totalEarnings: Double = 0.0,
    val totalDistance: Double = 0.0,  // miles
    val totalOnlineTime: Double = 0.0,  // hours
    val acceptanceRate: Double = 100.0,  // 0-100%
    val completionRate: Double = 100.0,  // 0-100%
    val onTimeRate: Double = 100.0,  // 0-100%
    val weeklyDeliveries: Int = 0,
    val weeklyEarnings: Double = 0.0,
    val weeklyHours: Double = 0.0,
    val weeklyDistance: Double = 0.0,
    val weekStartDate: Long = 0
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "rating" to rating,
        "totalDeliveries" to totalDeliveries,
        "completedDeliveries" to completedDeliveries,
        "cancelledDeliveries" to cancelledDeliveries,
        "totalEarnings" to totalEarnings,
        "totalDistance" to totalDistance,
        "totalOnlineTime" to totalOnlineTime,
        "acceptanceRate" to acceptanceRate,
        "completionRate" to completionRate,
        "onTimeRate" to onTimeRate,
        "weeklyDeliveries" to weeklyDeliveries,
        "weeklyEarnings" to weeklyEarnings,
        "weeklyHours" to weeklyHours,
        "weeklyDistance" to weeklyDistance,
        "weekStartDate" to weekStartDate
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): DriverStats {
            return DriverStats(
                rating = (data["rating"] as? Number)?.toDouble() ?: 5.0,
                totalDeliveries = (data["totalDeliveries"] as? Number)?.toInt() ?: 0,
                completedDeliveries = (data["completedDeliveries"] as? Number)?.toInt() ?: 0,
                cancelledDeliveries = (data["cancelledDeliveries"] as? Number)?.toInt() ?: 0,
                totalEarnings = (data["totalEarnings"] as? Number)?.toDouble() ?: 0.0,
                totalDistance = (data["totalDistance"] as? Number)?.toDouble() ?: 0.0,
                totalOnlineTime = (data["totalOnlineTime"] as? Number)?.toDouble() ?: 0.0,
                acceptanceRate = (data["acceptanceRate"] as? Number)?.toDouble() ?: 100.0,
                completionRate = (data["completionRate"] as? Number)?.toDouble() ?: 100.0,
                onTimeRate = (data["onTimeRate"] as? Number)?.toDouble() ?: 100.0,
                weeklyDeliveries = (data["weeklyDeliveries"] as? Number)?.toInt() ?: 0,
                weeklyEarnings = (data["weeklyEarnings"] as? Number)?.toDouble() ?: 0.0,
                weeklyHours = (data["weeklyHours"] as? Number)?.toDouble() ?: 0.0,
                weeklyDistance = (data["weeklyDistance"] as? Number)?.toDouble() ?: 0.0,
                weekStartDate = (data["weekStartDate"] as? Number)?.toLong() ?: 0
            )
        }
    }
}

data class DriverPreferences(
    val maxDeliveryDistance: Double = 10.0,  // miles
    val preferredAreas: List<String> = emptyList(),  // Zip codes or neighborhoods
    val acceptCashOrders: Boolean = true,
    val notificationsEnabled: Boolean = true,
    val soundEnabled: Boolean = true,
    val autoAcceptOrders: Boolean = false,
    val preferredShifts: List<String> = emptyList()  // "morning", "afternoon", "evening", "night"
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "maxDeliveryDistance" to maxDeliveryDistance,
        "preferredAreas" to preferredAreas,
        "acceptCashOrders" to acceptCashOrders,
        "notificationsEnabled" to notificationsEnabled,
        "soundEnabled" to soundEnabled,
        "autoAcceptOrders" to autoAcceptOrders,
        "preferredShifts" to preferredShifts
    )

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(data: Map<String, Any?>): DriverPreferences {
            return DriverPreferences(
                maxDeliveryDistance = (data["maxDeliveryDistance"] as? Number)?.toDouble() ?: 10.0,
                preferredAreas = (data["preferredAreas"] as? List<String>) ?: emptyList(),
                acceptCashOrders = data["acceptCashOrders"] as? Boolean ?: true,
                notificationsEnabled = data["notificationsEnabled"] as? Boolean ?: true,
                soundEnabled = data["soundEnabled"] as? Boolean ?: true,
                autoAcceptOrders = data["autoAcceptOrders"] as? Boolean ?: false,
                preferredShifts = (data["preferredShifts"] as? List<String>) ?: emptyList()
            )
        }
    }
}

File: apps/ios/eatfair-ios-shared/Sources/EatFairShared/Models/Driver.swift
import Foundation
import FirebaseFirestore

public struct Driver: Identifiable, Codable {
    @DocumentID public var id: String?

    // MARK: - Personal Information
    public var name: String
    public var email: String
    public var phone: String
    public var profileImageUrl: String?
    public var dateOfBirth: Int64?
    public var ssn4: String? // Last 4 digits only for verification

    // MARK: - Address
    public var address: DriverAddress?

    // MARK: - Driver's License
    public var driversLicense: DriversLicense?

    // MARK: - Vehicle Information
    public var vehicle: VehicleInfo?
    public var vehicleType: String // Car, Bike, Scooter, Motorcycle, Truck
    public var licensePlate: String

    // MARK: - Insurance
    public var insurance: InsuranceInfo?

    // MARK: - Bank/Payment Information
    public var bankAccount: BankAccountInfo?

    // MARK: - Status & Location
    public var isOnline: Bool
    public var isApproved: Bool
    public var approvalStatus: String // "pending", "approved", "rejected", "suspended"
    public var currentLatitude: Double
    public var currentLongitude: Double
    public var lastActive: Int64?
    public var currentSessionId: String?

    // MARK: - Stats (embedded for quick access)
    public var stats: DriverStats?

    // MARK: - Preferences
    public var preferences: DriverPreferences?

    // MARK: - Metadata
    public var createdAt: Int64
    public var updatedAt: Int64?
    public var onboardingCompletedAt: Int64?

    // MARK: - Background Check
    public var backgroundCheckStatus: String? // "pending", "passed", "failed"
    public var backgroundCheckDate: Int64?

    public init(
        id: String? = nil,
        name: String = "",
        email: String = "",
        phone: String = "",
        profileImageUrl: String? = nil,
        dateOfBirth: Int64? = nil,
        ssn4: String? = nil,
        address: DriverAddress? = nil,
        driversLicense: DriversLicense? = nil,
        vehicle: VehicleInfo? = nil,
        vehicleType: String = "Car",
        licensePlate: String = "",
        insurance: InsuranceInfo? = nil,
        bankAccount: BankAccountInfo? = nil,
        isOnline: Bool = false,
        isApproved: Bool = false,
        approvalStatus: String = "pending",
        currentLatitude: Double = 0.0,
        currentLongitude: Double = 0.0,
        lastActive: Int64? = nil,
        currentSessionId: String? = nil,
        stats: DriverStats? = nil,
        preferences: DriverPreferences? = nil,
        createdAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        updatedAt: Int64? = nil,
        onboardingCompletedAt: Int64? = nil,
        backgroundCheckStatus: String? = nil,
        backgroundCheckDate: Int64? = nil
    ) {
        self.id = id
        self.name = name
        self.email = email
        self.phone = phone
        self.profileImageUrl = profileImageUrl
        self.dateOfBirth = dateOfBirth
        self.ssn4 = ssn4
        self.address = address
        self.driversLicense = driversLicense
        self.vehicle = vehicle
        self.vehicleType = vehicleType
        self.licensePlate = licensePlate
        self.insurance = insurance
        self.bankAccount = bankAccount
        self.isOnline = isOnline
        self.isApproved = isApproved
        self.approvalStatus = approvalStatus
        self.currentLatitude = currentLatitude
        self.currentLongitude = currentLongitude
        self.lastActive = lastActive
        self.currentSessionId = currentSessionId
        self.stats = stats
        self.preferences = preferences
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.onboardingCompletedAt = onboardingCompletedAt
        self.backgroundCheckStatus = backgroundCheckStatus
        self.backgroundCheckDate = backgroundCheckDate
    }
}

// MARK: - Driver Address
public struct DriverAddress: Codable {
    public var street: String
    public var unit: String?
    public var city: String
    public var state: String
    public var zipCode: String
    public var country: String

    public init(
        street: String = "",
        unit: String? = nil,
        city: String = "",
        state: String = "",
        zipCode: String = "",
        country: String = "USA"
    ) {
        self.street = street
        self.unit = unit
        self.city = city
        self.state = state
        self.zipCode = zipCode
        self.country = country
    }

    public var fullAddress: String {
        var parts = [street]
        if let unit = unit, !unit.isEmpty { parts.append("Unit \(unit)") }
        parts.append("\(city), \(state) \(zipCode)")
        return parts.joined(separator: ", ")
    }
}

// MARK: - Driver's License
public struct DriversLicense: Codable {
    public var licenseNumber: String
    public var state: String // Issuing state
    public var expirationDate: Int64
    public var licenseClass: String // "A", "B", "C", "D", "M"
    public var frontImageUrl: String?
    public var backImageUrl: String?
    public var isVerified: Bool
    public var verifiedAt: Int64?

    public init(
        licenseNumber: String = "",
        state: String = "",
        expirationDate: Int64 = Int64(Date().addingTimeInterval(365*24*60*60).timeIntervalSince1970 * 1000),
        licenseClass: String = "C",
        frontImageUrl: String? = nil,
        backImageUrl: String? = nil,
        isVerified: Bool = false,
        verifiedAt: Int64? = nil
    ) {
        self.licenseNumber = licenseNumber
        self.state = state
        self.expirationDate = expirationDate
        self.licenseClass = licenseClass
        self.frontImageUrl = frontImageUrl
        self.backImageUrl = backImageUrl
        self.isVerified = isVerified
        self.verifiedAt = verifiedAt
    }

    public var isExpired: Bool {
        return expirationDate < Int64(Date().timeIntervalSince1970 * 1000)
    }

    public var expirationDateFormatted: String {
        let date = Date(timeIntervalSince1970: TimeInterval(expirationDate / 1000))
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        return formatter.string(from: date)
    }
}

// MARK: - Vehicle Information
public struct VehicleInfo: Codable {
    public var make: String // Toyota, Honda, Ford
    public var model: String // Camry, Civic, F-150
    public var year: Int
    public var color: String
    public var licensePlate: String
    public var vehicleType: String // Sedan, SUV, Truck, Motorcycle, Bicycle, Scooter

    // Registration
    public var registrationNumber: String?
    public var registrationExpirationDate: Int64?
    public var registrationImageUrl: String?

    public var isVerified: Bool

    public init(
        make: String = "",
        model: String = "",
        year: Int = Calendar.current.component(.year, from: Date()),
        color: String = "",
        licensePlate: String = "",
        vehicleType: String = "Sedan",
        registrationNumber: String? = nil,
        registrationExpirationDate: Int64? = nil,
        registrationImageUrl: String? = nil,
        isVerified: Bool = false
    ) {
        self.make = make
        self.model = model
        self.year = year
        self.color = color
        self.licensePlate = licensePlate
        self.vehicleType = vehicleType
        self.registrationNumber = registrationNumber
        self.registrationExpirationDate = registrationExpirationDate
        self.registrationImageUrl = registrationImageUrl
        self.isVerified = isVerified
    }

    public var displayName: String {
        return "\(year) \(make) \(model)"
    }
}

// MARK: - Insurance Information
public struct InsuranceInfo: Codable {
    public var provider: String // State Farm, Geico, Progressive
    public var policyNumber: String
    public var expirationDate: Int64
    public var coverageType: String // "liability", "full_coverage"
    public var insuranceCardImageUrl: String?
    public var isVerified: Bool

    public init(
        provider: String = "",
        policyNumber: String = "",
        expirationDate: Int64 = Int64(Date().addingTimeInterval(180*24*60*60).timeIntervalSince1970 * 1000),
        coverageType: String = "full_coverage",
        insuranceCardImageUrl: String? = nil,
        isVerified: Bool = false
    ) {
        self.provider = provider
        self.policyNumber = policyNumber
        self.expirationDate = expirationDate
        self.coverageType = coverageType
        self.insuranceCardImageUrl = insuranceCardImageUrl
        self.isVerified = isVerified
    }

    public var isExpired: Bool {
        return expirationDate < Int64(Date().timeIntervalSince1970 * 1000)
    }
}

// MARK: - Bank Account Information
public struct BankAccountInfo: Codable {
    public var bankName: String
    public var accountHolderName: String
    public var accountType: String // "checking", "savings"
    public var routingNumber: String // Should be encrypted in production
    public var accountNumberLast4: String // Only store last 4 digits
    public var isVerified: Bool
    public var stripeConnectId: String? // For Stripe Connect payouts

    public init(
        bankName: String = "",
        accountHolderName: String = "",
        accountType: String = "checking",
        routingNumber: String = "",
        accountNumberLast4: String = "",
        isVerified: Bool = false,
        stripeConnectId: String? = nil
    ) {
        self.bankName = bankName
        self.accountHolderName = accountHolderName
        self.accountType = accountType
        self.routingNumber = routingNumber
        self.accountNumberLast4 = accountNumberLast4
        self.isVerified = isVerified
        self.stripeConnectId = stripeConnectId
    }
}

// MARK: - Driver Preferences
public struct DriverPreferences: Codable {
    public var maxDeliveryDistance: Double // miles
    public var preferredAreas: [String] // Zip codes or neighborhoods
    public var acceptCashOrders: Bool
    public var notificationsEnabled: Bool
    public var soundEnabled: Bool
    public var autoAcceptOrders: Bool
    public var preferredShifts: [String] // "morning", "afternoon", "evening", "night"

    public init(
        maxDeliveryDistance: Double = 10.0,
        preferredAreas: [String] = [],
        acceptCashOrders: Bool = true,
        notificationsEnabled: Bool = true,
        soundEnabled: Bool = true,
        autoAcceptOrders: Bool = false,
        preferredShifts: [String] = ["morning", "afternoon", "evening"]
    ) {
        self.maxDeliveryDistance = maxDeliveryDistance
        self.preferredAreas = preferredAreas
        self.acceptCashOrders = acceptCashOrders
        self.notificationsEnabled = notificationsEnabled
        self.soundEnabled = soundEnabled
        self.autoAcceptOrders = autoAcceptOrders
        self.preferredShifts = preferredShifts
    }
}
